# Preparing data sets {#dataset}
## Case I: 10x data (Stewart et al., Nat Cancer, 2020)
```{r, eval = FALSE}
source("R/function_general.R")
```

The data were obtained from NCBI repository with accession number [GSE138474](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138474).
From the above, we downloaded scRNA-seq data with sample accessions numbers
GSM4104164 (for MDA-SC68 with vehicle treatment) and
GSM4104163 (for MDA-SC68 with cisplatin treatment).

The raw count matrices were produced by 10x cellranger with a reference
genome GRCh38-3.0.0.
The following function `read_matrix_10xdata()` and `read_gene_10xdata()`
process the dumped files into a raw count matrix and gene dataframe.
In this function, `make.unique()` in `base` package is applied for naming gene
symbols, which appends a sequential number with a period delimiter for every
repeat name encountered.
```{r, eval = FALSE}
read_matrix_10xdata <- function(path_dir){
  barcode.path <- paste0(path_dir, "barcodes.tsv.gz")
  feature.path <- paste0(path_dir, "features.tsv.gz")
  matrix.path  <- paste0(path_dir, "matrix.mtx.gz")
  mat <- as.matrix(readMM(file = matrix.path))
  genes <- read.delim(feature.path, header = FALSE, stringsAsFactors = FALSE)
  barcodes <- read.delim(barcode.path, header = FALSE, stringsAsFactors = FALSE)
  rownames(mat) <- make.unique(as.character(genes$V2))
  colnames(mat) <- barcodes$V1

  return(mat)
}

read_gene_10xdata <- function(path_dir){
  feature.path <- paste0(path_dir, "features.tsv.gz")
  genes <- read.delim(feature.path, header = FALSE, stringsAsFactors = FALSE)
  
  return(genes)
}
```

The following function `make_asurat_obj()` creates an ASURAT object.
Note that the first argument `mat` must have row and column names
as gene symbols and cell barcodes, respectively.
```{r, eval = FALSE}
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
path_dir1 <- "data/2020_001_stewart/sc68_vehi/SRR10211594_count/"
path_dir1 <- paste(path_dir1, "filtered_feature_bc_matrix/", sep = "")
sc68_vehi <- make_asurat_obj(mat = read_matrix_10xdata(path_dir = path_dir1),
                             obj_name = "sc68_vehi")
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
path_dir2 <- "data/2020_001_stewart/sc68_cisp/SRR10211593_count/"
path_dir2 <- paste(path_dir2, "filtered_feature_bc_matrix/", sep = "")
sc68_cisp <- make_asurat_obj(mat = read_matrix_10xdata(path_dir = path_dir2),
                             obj_name = "sc68_cisp")
```
Users can see `obj[["variable"]]` is a data frame of gene symbol and
Entrez Gene ID.

Users can add Ensembl IDs to the data frame as follows (this step is optional).
```{r, eval = FALSE}
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
sc68_vehi[["variable"]][["ensembl"]] <-
  read_gene_10xdata(path_dir = path_dir1)$V1
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
sc68_cisp[["variable"]][["ensembl"]] <-
  read_gene_10xdata(path_dir = path_dir2)$V1
```

In what follows, we convert gene symbols to Entrez Gene IDs using
`org.Hs.eg.db` package.
Although users can customize this step, at least
`obj[["variable"]][["entrez"]]` must be prepared, which can include `NA`.
```{r, eval = FALSE}
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = sc68_vehi[["variable"]][["symbol"]],
                                    columns = "ENTREZID", keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
sc68_vehi[["variable"]][["entrez"]] <- dictionary$ENTREZID
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = sc68_cisp[["variable"]][["symbol"]],
                                    columns = "ENTREZID", keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
sc68_cisp[["variable"]][["entrez"]] <- dictionary$ENTREZID
```

Save the objects.
```{r, eval = FALSE}
saveRDS(sc68_vehi, file = "backup/01_001_sc68_vehi_raw.rds")
saveRDS(sc68_cisp, file = "backup/02_001_sc68_cisp_raw.rds")
```



## Case II: PDAC-A ST and inDrop (Moncada et al., Nat Biotechnol, 2020)
The data were obtained from NCBI repository with accession number [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672).



### PDAC-A ST
From [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672),
we downloaded spatial transcriptome data with sample accession number
GSM3036911 (for PDAC-A ST).

```{r, eval = FALSE}
source("R/function_general.R")
source("R/Spaniel/R/spanielPlot.R")
source("R/Spaniel/R/utilities.R")
source("R/Spaniel/R/spanielPlotInternals.R")
```

Load read count tables and crate ASURAT objects.
In order to create ASURAT objects, read count tables `mat` must have row and
column names as gene symbols and cell barcodes, respectively.
Hence, we first change the row names of `mat` into gene symbols using
`org.Hs.eg.db`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_ast1
# ----------------------------------------
fn <- "data/2020_001_Moncada/pdac_ast1/SRR6825057_stdata.tsv"
mat <- read.table(fn, header = TRUE, stringsAsFactors = FALSE, row.names = 1)
mat <- t(mat)
ensembl <- rownames(mat)
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = ensembl,
                                    columns = c("SYMBOL", "ENTREZID"),
                                    keytype = "ENSEMBL")
dictionary <- dictionary[!duplicated(dictionary$ENSEMBL), ]
dictionary[which(is.na(dictionary$SYMBOL)),]$SYMBOL <- as.character("NA")
rownames(mat) <- make.unique(as.character(dictionary$SYMBOL))
pdac_ast1 <- make_asurat_obj(mat = mat, obj_name = "pdac_ast1")
```

In addition, load Seurat objects with spatial coordinate information,
which was obtained from the author of
[SPOTlight](https://academic.oup.com/nar/article/49/9/e50/6129341).
```{r, eval = FALSE}
fn <- "data/2020_001_Moncada/MarcElosuaBayes/processed_data/PDAC-A_ST_list.RDS"
pdac_list <- readRDS(file = fn)
pdac_ast1_seurat <- pdac_list$GSM3036911
```

Add `pdac_ast1_seurat@images` to `pdac_ast1[["misc"]][["images"]]`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_ast1
# ----------------------------------------
pdac_ast1[["misc"]][["images"]] <- pdac_ast1_seurat@images
```

Since the above `pdac_ast1` includes spatial coordinates outside of tissues,
remove such spots by comparing with `pdac_ast1_seurat@meta.data[["spot"]]`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_ast1
# ----------------------------------------
tmp <- pdac_ast1[["sample"]][["barcode"]]
for(i in 1:length(tmp)){
 x <- as.integer(strsplit(tmp[i], "x")[[1]][1])
 y <- as.integer(strsplit(tmp[i], "x")[[1]][2])
 pdac_ast1[["sample"]][["spot"]][i] <- paste("PDAC-A-ST1_", x, "x", y, sep = "")
 pdac_ast1[["sample"]][["x"]][i] <- x
 pdac_ast1[["sample"]][["y"]][i] <- y
}
tmp <- pdac_ast1[["sample"]]
ref <- pdac_ast1_seurat@meta.data[["spot"]]
pdac_ast1[["sample"]] <- tmp[which(tmp$spot %in% ref),]

mat <- pdac_ast1[["data"]][["raw"]]
mat <- mat[, which(colnames(mat) %in% pdac_ast1[["sample"]][["barcode"]])]
pdac_ast1[["data"]][["raw"]] <- mat
```

In order to use `spanielPlot()` in `Spaniel` package, we temporarily convert
ASURAT objects into Seurat objects and visualize the number of reads
of a given gene at the plots on PDAC tissue.
```{r, eval = FALSE}
fn <- "data/2020_001_Moncada/MarcElosuaBayes/pre-processed_data/"
fn <- paste0(fn, "PDAC-A/GSM3036911/spatial/")
fn <- paste0(fn, "tissue_positions_list_spatial_object.tsv")
mat <- pdac_ast1[["data"]][["raw"]]
colnames(mat) <- pdac_ast1[["sample"]][["spot"]]  # This is required by Spaniel
se <- Spaniel::createSeurat(counts = mat, barcodeFile = fn,
                            projectName = "PDAC-A", sectionNumber = "1")
se@images <- pdac_ast1_seurat@images
se <- Seurat::ScaleData(se)
p <- spanielPlot_2(object = se, grob = se@images[[1]], plotType = "Gene",
                   gene = "TM4SF1", ptSizeMax = 2,  ptSizeMin = 0)
filename <- "figures/figure_07_0001.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.6, height = 4.1)
```

<img src="figures/figure_07_0001.png" width="400px">

In what follows, we convert Ensembl IDs to Entrez Gene IDs.
Note that at least `obj[["variable"]][["entrez"]]` must be prepared,
which can include `NA`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_ast1
# ----------------------------------------
pdac_ast1[["variable"]][["entrez"]] <- dictionary$ENTREZID
pdac_ast1[["variable"]][["ensembl"]] <- dictionary$ENSEMBL
```

Save the objects.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_ast1
# ----------------------------------------
saveRDS(pdac_ast1, file = "backup/05_001_pdac_ast1_raw.rds")
```



### PDAC-A inDrop
From [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672),
we downloaded spatial transcriptome data with sample accession numbers
GSM3036909, GSM3036910, GSM3405527, GSM3405528, GSM3405529, and GSM3405530
(for PDAC-A inDrop1-inDrop6).

```{r, eval = FALSE}
source("R/plot.R")
source("R/function_general.R")
```

Load read count tables and create ASURAT objects.
Here we create a list of ASURAT objects.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
fn <- c("data/2020_001_Moncada/pdac_arna/PDACA_1/results/gene_expression.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_2/results/gene_expression.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_3/results/gene_expression.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_4/results/gene_expression.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_5/results/gene_expression.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_6/results/gene_expression.tsv")
pdac_arna <- list()
for(i in 1:length(fn)){
  mat <- read.table(fn[i], header = T, stringsAsFactors = F, row.names = 1)
  pdac_arna[[i]] <- make_asurat_obj(mat = mat,
                                    obj_name = paste("pdac_arna_", i, sep = ""))
}
```
Users can see `obj[["variable"]]` is a data frame of gene symbol and
Entrez Gene ID.
Note that we selected only protein coding genes for PDAC-A inDrop datasets.

Users can add Ensembl IDs to the data frame as follows (this step is optional).
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
fn <- c("data/2020_001_Moncada/pdac_arna/PDACA_1/results/gene_names.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_2/results/gene_names.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_3/results/gene_names.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_4/results/gene_names.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_5/results/gene_names.tsv",
        "data/2020_001_Moncada/pdac_arna/PDACA_6/results/gene_names.tsv")
genes <- list()
for(i in 1:length(fn)){
  genes[[i]] <- read.delim(fn[i], header = TRUE)
}
for(i in 1:(length(fn)-1)){
  print(identical(genes[[i]], genes[[i+1]]))
}
genes <- genes[[1]]
for(i in 1:length(fn)){
  if(identical(genes$Gene.name, pdac_arna[[i]][["variable"]][["symbol"]])){
    pdac_arna[[i]][["variable"]][["ensembl"]] <- genes$Gene.ID
  }else{
    stop("Please add Ensembl IDs in other ways.")
  }
}
```

In what follows, we convert Ensembl IDs to Entrez Gene IDs using
`org.Hs.eg.db` package.
Although users can customize this step, at least
`obj[["variable"]][["entrez"]]` must be prepared, which can include `NA`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = genes$Gene.ID,
                                    columns = "ENTREZID", keytype = "ENSEMBL")
dictionary <- dictionary[!duplicated(dictionary$ENSEMBL), ]
for(i in 1:length(fn)){
  pdac_arna[[i]][["variable"]][["entrez"]] <- dictionary$ENTREZID
}
```

Let us briefly check the data qualities.
The following function `do_quickQC()` counts the number of reads,
the number of non-zero expressing genes, and the percent of reads
that map to the mitochondrial genes, e.g., `MT-Co1`, `mt-Co1`, etc.
We assume that function is usually used to remove genes for which the number
of non-zero expressing cells is less than `min_nsamples`,
but here we must set `min_nsamples = 0`.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
for(i in 1:length(fn)){
  pdac_arna[[i]] <- do_quickQC(obj = pdac_arna[[i]], min_nsamples = 0,
                               mitochondria_symbol = "^MT-")
}
```

The following function `plot_metaData()` outputs a scatter plot between the
input meta data for each cell
The arguments are `df` (two-dimensional data frame),
`title` (title of the plot), `xlabel` (x-axis label), and
`ylabel` (y-axis label).
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
my_func <- function(obj, metaname1, metaname2){
  df <- data.frame(x = log(obj[["sample"]][[metaname1]] + 1),
                   y = log(obj[["sample"]][[metaname2]] + 1))
  p <- plot_metaData(df = df,
                    title = obj[["history"]][["make_asurat_obj"]][["obj_name"]],
                    title_size = 20, xlabel = "log(nReads + 1)",
                    ylabel = "log(nGenes + 1)")
  return(p)
}

for(i in 1:length(fn)){
  p <- my_func(obj = pdac_arna[[i]], metaname1 = "nReads", metaname2 = "nGenes")
  filename <- paste("figures/figure_06_", sprintf("%04d", i), ".png", sep = "")
  ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)
}
```

<img src="figures/figure_06_0001.png" width="230px">
<img src="figures/figure_06_0002.png" width="230px">
<img src="figures/figure_06_0003.png" width="230px">

<img src="figures/figure_06_0004.png" width="230px">
<img src="figures/figure_06_0005.png" width="230px">
<img src="figures/figure_06_0006.png" width="230px">

The following function `trim_samples()` filters out the low quality
cells by setting minimum and maximum values of the calculated QC metrics.
This function will be used when we seriously control the data quality,
but here we only remove cells with very small number of reads
by setting `min_nReads = 100` while other parameters ineffective.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
for(i in 1:length(fn)){
  pdac_arna[[i]] <- trim_samples(obj = pdac_arna[[i]],
                                 min_nReads = 100, max_nReads = 1e+10,
                                 min_nGenes = 0, max_nGenes = 1e+10,
                                 min_percent_MT = 0, max_percent_MT = 100)
}
```

Based on the above figures, we easily concatenate all the read count tables
and recreate an ASURAT object as follows:
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
variable <- pdac_arna[[1]][["variable"]]
mat <- pdac_arna[[1]][["data"]][["raw"]]
for(i in 2:length(fn)){
  mat <- cbind(mat, pdac_arna[[i]][["data"]][["raw"]])
}
sample <- make.unique(colnames(mat))
colnames(mat) <- sample
pdac_arna <- make_asurat_obj(mat = mat, obj_name = "pdac_arna")
pdac_arna[["variable"]] <- variable
```

Save the objects.
```{r, eval = FALSE}
# ----------------------------------------
# pdac_arna
# ----------------------------------------
saveRDS(pdac_arna, file = "backup/06_001_pdac_arna_raw.rds")
```



## Case III: input 10x PBMC data
```{r, eval = FALSE}
source("R/function_general.R")
```

The data (peripheral blood mononuclear cells from a healthy donor) were
obtained from 10x Genomics repository (PBMC 4k and 6k).

The following function `read_matrix_10xdata_v2()` and `read_gene_10xdata()`
process the fastq files into a raw count matrix and gene dataframe.
In this function, `make.unique()` in `base` package is applied for naming gene
symbols, which appends a sequential number with a period delimiter for every
repeat name encountered.
```{r, eval = FALSE}
read_matrix_10xdata_v2 <- function(path_dir){
  library(Seurat)
  obj <- Read10X(data.dir = path_dir, gene.column = 2, unique.features = TRUE,
                 strip.suffix = FALSE)
  obj <- CreateSeuratObject(counts = obj)
  mat <- as.matrix(obj@assays[["RNA"]]@counts)
  return(mat)
}

read_gene_10xdata <- function(path_dir){
  feature.path <- paste0(path_dir, "genes.tsv")
  genes <- read.delim(feature.path, header = FALSE, stringsAsFactors = FALSE)
  
  return(genes)
}
```

The following function `make_asurat_obj()` creates an ASURAT object.
Note that the first argument `mat` must have row and column names
as gene symbols and cell barcodes, respectively.
```{r, eval = FALSE}
# ----------------------------------------
# pbmc_4000
# ----------------------------------------
path_dir1 <- "data/2020_001_10xgenomics/pbmc_4000/"
path_dir1 <- paste(path_dir1, "filtered_gene_bc_matrices/GRCh38/", sep = "")
pbmc_4000 <- make_asurat_obj(mat = read_matrix_10xdata_v2(path_dir = path_dir1),
                             obj_name = "pbmc_4000")
# ----------------------------------------
# pbmc_6000
# ----------------------------------------
path_dir2 <- "data/2020_001_10xgenomics/pbmc_6000/"
path_dir2 <- paste(path_dir2, "filtered_matrices_mex/hg19/", sep = "")
pbmc_6000 <- make_asurat_obj(mat = read_matrix_10xdata_v2(path_dir = path_dir2),
                             obj_name = "pbmc_6000")
```
Users can see `obj[["variable"]]` is a data frame of gene symbol and
Entrez Gene ID.

Users can add Ensembl IDs to the data frame as follows (this step is optional).
```{r, eval = FALSE}
# ----------------------------------------
# pbmc_4000
# ----------------------------------------
pbmc_4000[["variable"]][["ensembl"]] <-
  read_gene_10xdata(path_dir = path_dir1)$V1
# ----------------------------------------
# pbmc_6000
# ----------------------------------------
pbmc_6000[["variable"]][["ensembl"]] <-
  read_gene_10xdata(path_dir = path_dir2)$V1
```

In what follows, we convert gene symbols to Entrez Gene IDs using
`org.Hs.eg.db` package.
Although users can customize this step, at least
`obj[["variable"]][["entrez"]]` must be prepared, which can include `NA`.
```{r, eval = FALSE}
# ----------------------------------------
# pbmc_4000
# ----------------------------------------
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = pbmc_4000[["variable"]][["symbol"]],
                                    columns = "ENTREZID", keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
pbmc_4000[["variable"]][["entrez"]] <- dictionary$ENTREZID
# ----------------------------------------
# pbmc_6000
# ----------------------------------------
dictionary <- AnnotationDbi::select(org.Hs.eg.db,
                                    key = pbmc_6000[["variable"]][["symbol"]],
                                    columns = "ENTREZID", keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
pbmc_6000[["variable"]][["entrez"]] <- dictionary$ENTREZID
```

Save the objects.
```{r, eval = FALSE}
saveRDS(pbmc_4000, file = "backup/10_001_pbmc_4000_raw.rds")
saveRDS(pbmc_6000, file = "backup/11_001_pbmc_6000_raw.rds")
```
